---
title: "Árvores nativas"
author: "Heron Hilário"
date: "5/11/2022"
output: 
  html_document:
    code_download: yes
    theme: flatly
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
{
  library(Biostrings)
library(DECIPHER)
library(dplyr)
library(stringr)
library(ggtree)
library(ape)
  }
```

# Fazendo árvores a partir de sequências

```{r,echo=TRUE, eval=FALSE}

# Read DNA from fastas ----

## A função readDNAStringSet() do pacote Biostrings é chamada para criar um objeto 
## para armazenar as sequências de DNA do DB (objeto db_fasta) e das ASVs 
## identificadas no pipeline (asvs_fasta). A flag RemoveGaps() do pacote DECIPHER
## é utilizada para remover os gaps ("-" ou ".") que estejam nas sequências.

## As many as you want (for the same region!)
# dna_fasta_x <- Biostrings::readDNAStringSet(filepath = "/home/gabriel/projetos/db-LGC/data/mai22/LGC12Sdb_252seqs-mai22-order_clust.fasta") %>%
#   DECIPHER::RemoveGaps()

{
## Read DB

db_fasta <- Biostrings::readDNAStringSet(filepath = "/home/heron/prjcts/fish_eDNA/DB/mai22/DB/LGC12Sdb_251seqs-mai22-pretty_names_noGaps.fasta") %>%
  DECIPHER::RemoveGaps()

## Read ASVs 
asvs_fasta <- Biostrings::readDNAStringSet(filepath = "/home/gabriel/projetos/peixes-eDNA/analises/pipeline-modelo/results/pipeline-modelo-all_ASVs_all_primers.fasta") %>%
  DECIPHER::RemoveGaps()
}

# Visualizar os objetos com as seqs ----

## Abrir a visualização das sequências armazenadas nos objetos db_fasta e asvs_fasta

{
  BrowseSeqs(db_fasta)
  BrowseSeqs(asvs_fasta)
  }

# Selecionar a região do DB pertinente às ASVs ----

## Alinhar as seqs do DB

db_fasta_algn <- DECIPHER::AlignSeqs(myXStringSet = db_fasta, 
                                      refinements = 100,
                                      iterations = 100,
                                      verbose = TRUE)

BrowseSeqs(db_fasta_algn)

## Encontrar o primer nas seqs

## Encontrar a região do primer reverse nas seqs,
## deve ser feito manualmente :/ Ctrl+F no browser
## Identify informative region (EX: amplicon)

## Select MiFish Amplicon
#FWD = GTCGGTAAAACTCGTGCCAGC 
#REV = CATAGTGGGGTATCTAATCCCAGTTTG  
  DNAStringSet("CATAGTGGGGTATCTAATCCCAGTTTG") %>% 
  reverseComplement()
#REV (rev comp) = CAAACTGGGATTAGATACCCCACTATG

## Based on your visualization, select only the region common to all species
DECIPHER::BrowseSeqs(db_fasta_algn)

## Procurar pela sequência acima no alinhamento dna_fasta_algn e definir
## qual a região a ser trimada. Percebemos que é no nt 251

# Cut your alignment 
dna_fasta_algn_cut <- Biostrings::subseq(x = db_fasta_algn,start = 1, end = 251) 

# Combinar os arquivos e sequências requeridas ----

## dna_fasta <- c(dna_fasta_1,dna_fasta_2)

{
  dna_fasta <- c(dna_fasta_algn_cut, asvs_fasta) %>% unique() %>% RemoveGaps()

## Check DNA seqs 
dna_fasta

## Check their name 
BrowseSeqs(dna_fasta)
  }

# Alterar os nomes (optional) ----

## all together

  # dna_fasta <- names(dna_fasta) %>% str_remove(pattern = "SF \\| |JQ \\| ")

## Only some (two ways of subsetting)

  # names(dna_fasta)[10]
# names(dna_fasta)[10] <- "P.costatus"

  # names(dna_fasta[10])
# names(dna_fasta[10]) <- "P.costatus2"

# names(dna_fasta[10]) <- 
  # names(dna_fasta[10]) %>% str_replace(pattern = "Prochilodus",
                                   # replacement = "P.")

# Align your seqs ----

dna_fasta_algn <- DECIPHER::AlignSeqs(myXStringSet = dna_fasta, 
                                      refinements = 100,
                                      iterations = 100,
                                      verbose = TRUE)
DECIPHER::BrowseSeqs(dna_fasta_algn)

## Save your alignment

Biostrings::writeXStringSet(x = dna_fasta_algn,
                            file =  "/home/gabriel/projetos/lagoa_ingleses/tree/asvs-LI_db.fasta")
}

## Remove some distant species or with very short seqs
## NOTE: cuztomize as you want
amplicon_algn <- amplicon_algn[!(names(amplicon_algn) %>% str_detect(pattern = "Out"))]

# Calculate distance matrix ----
amplicon_dist <- DECIPHER::DistanceMatrix(myXStringSet = amplicon_algn,
                                            # includeTerminalGaps = TRUE,
                                            includeTerminalGaps = FALSE,      #test both versions on the final output
                                            correction = "Jukes-Cantor",
                                            processors = 20,
                                            verbose = TRUE)

# Build tree ----

amplicon_tree <- ape::njs(amplicon_dist)

# Plot tree tow view ----
amplicon_tree_plot <-
  ggtree(tr = amplicon_tree) + 
  theme_tree2() +
  geom_tiplab(offset = 0,align = T) + 
  xlim(0, 2)

amplicon_tree_plot %>% ggsave(filename = "/home/heron/prjcts/misc/LGC/tree_12Sdb_test.pdf",
                              device = "pdf",width = 12,height = 40)


# Save tree ----
ape::write.tree(phy = dna_fasta_tree,file = "~/prjcts/fish_eDNAarvore/12SDB_MiFish_tree.nwk")

# Customize (manually)-----

## on MEGA software, drag and drop the .nwk file. Use tools to generate a pretty visualization
```
